packages/server/test/websockets.test.ts

```diff
@@ -0,0 +1,171 @@
+});
+
+test('subscriptionOnce()', async () => {
+  const ee = new EventEmitter();
+  type Message = {
+    id: string;
+  };
+  let ws: any = null;
+  const { client, close } = routerToServerAndClient(
+    trpc.router().subscription('onMessage', {
+      input: z.string().optional(),
+      resolve() {
+        ee.emit('server:connect');
+        return new trpc.Subscription<Message>({
+          start(emit) {
+            const onMessage = (data: Message) => {
+              emit.data(data);
+            };
+            ee.on('server:msg', onMessage);
+            return () => ee.off('server:msg', onMessage);
+          },
+        });
+      },
+    }),
+    {
+      client({ wssUrl }) {
+        ws = new WebSocket(wssUrl);
+        return {
+          links: [webSocketLink({ ws })],
+        };
+      },
+    },
+  );
+  ee.once('server:connect', () => {
+    setImmediate(() => {
+      ee.emit('server:msg', {
+        id: '1',
+      });
+    });
+  });
+  const msgs = await client.subscriptionOnce('onMessage', '');
+
+  expectTypeOf(msgs).toMatchTypeOf<Message[]>();
+  expect(msgs).toMatchInlineSnapshot(`
+    Object {
+      "id": "1",
+    }
+  `);
+
+  await waitFor(() => {
+    expect(ee.listenerCount('server:msg')).toBe(0);
+    expect(ee.listenerCount('server:error')).toBe(0);
+  });
+  close();
+  ws.close();
+});
+
+test('subscription()', async () => {
+  const ee = new EventEmitter();
+  type Message = {
+    id: string;
+  };
+  let ws: any = null;
+  const { client, close } = routerToServerAndClient(
+    trpc.router().subscription('onMessage', {
+      input: z.string().optional(),
+      resolve() {
+        ee.emit('server:connect');
+        return new trpc.Subscription<Message>({
+          start(emit) {
+            const onMessage = (data: Message) => {
+              emit.data(data);
+            };
+            ee.on('server:msg', onMessage);
+            return () => ee.off('server:msg', onMessage);
+          },
+        });
+      },
+    }),
+    {
+      client({ wssUrl }) {
+        ws = new WebSocket(wssUrl);
+        return {
+          links: [webSocketLink({ ws })],
+        };
+      },
+    },
+  );
+  ee.once('server:connect', () => {
+    setImmediate(() => {
+      ee.emit('server:msg', {
+        id: '1',
+      });
+      ee.emit('server:msg', {
+        id: '1',
+      });
+    });
+  });
+  const callback = jest.fn();
+  const $sub = client.$subscription('onMessage');
+
+  $sub.subscribe((value) => {
+    callback(value);
+  });
+
+  await waitFor(() => {
+    expect(callback).toHaveBeenCalledTimes(2);
+  });
+
+  expect(callback.mock.calls).toMatchInlineSnapshot(`
+    Array [
+      Array [
+        Object {
+          "data": Object {
+            "id": "1",
+          },
+          "ok": true,
+        },
+      ],
+      Array [
+        Object {
+          "data": Object {
+            "id": "1",
+          },
+          "ok": true,
+        },
+      ],
+    ]
+  `);
+
+  $sub.destroy();
```

Comment:
- you call `client.$subscription('x')` to return an observable
- then you subscribe to the observable
- then you call `$sub.destroy()`

i feel it's a bit convoluted, what you would think would be ideal, @gmencz?

Reply:
What is `onMessage`? Is it some topic you subscribe to and other clients can publish to it?

Reply:
Each of them would call the `client.subscribe(topic)` method and whenever a client publishes something on that topic, the server broadcasts to all subscribers of the topic, that is how https://github.com/FungiRealtime does it and it's been working pretty well